### PR TITLE
Show tiles instead of thumbnails in ImageSet detail view

### LINKED ISSUE #518: Enable tile serving in ImageSet detail view
See #516 and notice that we are currently sending a thumbnail preview to the web client for viewing in GeoJS. We need to switch this to serve tiles of the imagery (note that this is not in a geospatial context)

### PR DESCRIPTION
**Pull request by mvandenburgh**
Uses GeoJS's tile viewing capabilities to show tiles in the ImageSet detail view instead of a thumbnail.

Resolve #518

### CODE REVIEW

**Comment by banesullivan** [comment]
So far this is coming along nicely.  A few issue though:

- Images should not repeat
- Z layer issues
- Images do not start out in the center of the frame
- Some JPEG images' aspect ratio were off

| bad initial position | repeating images | incorrect aspect ratio |
|---|---|---
| Screen Shot 2021-09-17 at 9 52 52 AM |  Screen Shot 2021-09-17 at 9 50 45 AM |  Screen Shot 2021-09-17 at 9 50 31 AM |


Z layer issues: tiles are being requested and this error is thrown a lot:

```
django_1    | [17/Sep/2021 10:50:07] "GET /api/image_process/imagery/55/tiles/2/2/3.png?band=0 HTTP/1.1" 500 104722
django_1    | Internal Server Error: /api/image_process/imagery/55/tiles/2/1/3.png
django_1    | Traceback (most recent call last):
django_1    |   File "/usr/local/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
django_1    |     response = get_response(request)
django_1    |   File "/usr/local/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
django_1    |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
django_1    |   File "/usr/local/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
django_1    |     return view_func(*args, **kwargs)
django_1    |   File "/usr/local/lib/python3.8/site-packages/django/views/generic/base.py", line 70, in view
django_1    |     return self.dispatch(request, *args, **kwargs)
django_1    |   File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
django_1    |     response = self.handle_exception(exc)
django_1    |   File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
django_1    |     self.raise_uncaught_exception(exc)
django_1    |   File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
django_1    |     raise exc
django_1    |   File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 506, in dispatch
django_1    |     response = handler(request, *args, **kwargs)
django_1    |   File "/opt/django-project/django-rgd-imagery/rgd_imagery/rest/tiles.py", line 50, in get
django_1    |     tile_binary = tile_source.getTile(x, y, z)
django_1    |   File "/usr/local/lib/python3.8/site-packages/large_image/cache_util/cache.py", line 102, in wrapper
django_1    |     v = func(self, *args, **kwargs)
django_1    |   File "/usr/local/lib/python3.8/site-packages/large_image_source_pil/__init__.py", line 172, in getTile
django_1    |     raise TileSourceException('z layer does not exist')
django_1    | large_image.exceptions.TileSourceException: z layer does not exist
```

**Comment by mvandenburgh** [comment]
> * Images should not repeat
> * Z layer issues
> * Images do not start out in the center of the frame
> * Some JPEG images' aspect ratio were off

These should be fixed now, @banesullivan.

**Review by banesullivan — APPROVED** [review]
This is looking and working great now!! Thanks, @mvandenburgh!

@manthey, if you have a chance, would you please look over the usage of GeoJS here?

File: django-rgd-imagery/rgd_imagery/templates/rgd_imagery/imageset_detail.html [hunk 1]
```diff
@@ -168,12 +101,41 @@
...
+        '#imageViewer', tileinfo.sizeX, tileinfo.sizeY, tileinfo.tileWidth, tileinfo.tileHeight);
+      params.layer.url = `${host}/api/image_process/imagery/${image_id}/tiles/{z}/{x}/{y}.png?band=${band}`;
+
+      const imageViewer = geo.map(params.map);
+
+      imageViewer.zoomRange({
+        // do not set a min limit so that bounds clamping determines min
+        min: -Infinity,
+        max: 12,
+      });
```
Comment (manthey) [inline django-rgd-imagery/rgd_imagery/templates/rgd_imagery/imageset_detail.html:124]: The `params.map` has some settings for the zoom range (`min` and `max` should be set to allow full navigation within the image, `zoom` is the starting zoom level and is set to the min value).  If you want to be able to zoom further, subtracting from the `params.map.min` and adding to `params.map.max` before calling `map(params.map)` should be sufficient.

**Comment by banesullivan** [comment]
I'm noticing some odd behavior with the tile requests coming from GeoJS in the ImageSet viewer.

Take the example of overhead raster imagery. When I view the image in world coordinates (the raster detail view page), I can zoom in to a Z-layer depth of 16+ and see individual pixels. However in the ImageSet viewer in image coordinates for the same image, I can only get to a Z-layer depth of 5 and the image looks overly smooth. 

(I'm zooming in a watching the request logs to get this max z-layer depth)

| Geospatial View | Image View |
|---|---|
| Screen Shot 2021-09-22 at 9 37 29 AM | Screen Shot 2021-09-22 at 9 37 33 AM| 
| Screen Shot 2021-09-22 at 9 38 51 AM | Screen Shot 2021-09-22 at 9 38 53 AM |

Tile request coming from GeoJS:
    - Raster with max Z layer of 16: `/api/image_process/imagery/27/tiles/16/17966/27060.png?projection=EPSG:3857&band=0`
    - Image with max Z layer of 3`/api/image_process/imagery/27/tiles/3/1/0.png?band=0` |

**Comment by manthey** [comment]
This might be right.

With a projection, the layers are referenced to the world (layer 0 is the whole world in one tile), so layer 16 probably makes sense. For non-projected viewer, the layers are related to the pixel size of the image (layer 0 is the whole image scaled to fit within one tile). so layer 3 *could* be the maximum layer (if, for instance, the tile size is 256x256 and the image size is less than 256*2**3).  Generally, we can allow zooming further than the image really warrants.  In a projection, GDAL renders the image to the projected space.  Without a projection, the image is rendered and then enlarged by the webgl texture interpolator.  

When GDAL reprojects data, the individual pixel data can conceptually be sited at a specific coordinates OR cover the area of spatial region.  If the first, you'll get some sort of smooth interpolation.  If the second, you'll get blocky pixel-looking data as you zoom in.  The webgl interpolator, on the other hand, always does interpolation, not nearest-neighbor (because this is *generally* what is desired unless you are trying to edit individual pixels).

**Comment by banesullivan** [comment]
Hm okay, thank you for the explanation @manthey!

**Pull request merged**